### PR TITLE
login: enable baseline SSO support for enterprise users

### PR DESCRIPTION
**Pull request by zkat**
This PR is meant to track progress on implementing the feature described over at #13388.

As of right now, it is able to perform a SAML-based login against @bcoe's related npme branch.

### CODE REVIEW

File: lib/adduser.js [hunk 1]
```diff
@@ -15,7 +11,7 @@ try {
 
 adduser.usage = usage(
   'adduser',
-  'npm adduser [--registry=url] [--scope=@orgname] [--always-auth]'
+  'npm adduser [--registry=url] [--scope=@orgname] [--auth-type=legacy] [--always-auth]'
```
Comment (watilde) [inline lib/adduser.js]: This new option should be added into the cli docs as well
https://github.com/npm/npm/blob/master/doc/cli/npm-adduser.md

Comment (zkat) [inline lib/adduser.js]: done! Thanks for the reminder!

**Comment by legodude17** [comment]
I would really like to have GitHub integration in `npm`. +100

File: doc/cli/npm-adduser.md [hunk 2]
```diff
@@ -66,6 +66,13 @@ different hostnames. See `always-auth` in `npm-config(7)` for more details on
 always-auth. Registry-specific configuration of `always-auth` takes precedence
 over any global configuration.
 
+### auth-type
+
+* Default: `'legacy'`
+* Type: `'legacy'`, `'sso'`, `'saml'`, `'oauth'`
+
+What authentication strategy to use with `adduser`/`login`.
```
Comment (iarna) [inline doc/cli/npm-adduser.md]: This description feels a little redundant given that this _is_ the `adduser`/`login` doc. Some description of under what circumstances you'd want to use the other strategies would probably be helpful, even if it's mostly a pointer to, like, npme docs.

File: lib/auth/sso.js [hunk 3]
```diff
@@ -0,0 +1,59 @@
...
+var npm = require('../npm.js')
+var output = require('../utils/output')
+var opener = require('opener')
+
+module.exports.login = function login (creds, registry, scope, cb) {
+  var ssoType = npm.config.get('sso-type')
+  if (!ssoType) { return cb(new Error('Missing option: sso-type')) }
+
+  var params = {
+    // We're reusing the legacy login endpoint, so we need some dummy
```
Comment (iarna) [inline lib/auth/sso.js]: Oh huh, I hadn't realized this is how this got squeezed in!

**Review by iarna — APPROVED** [review]
Looks good, with maybe a touch more documentation let's land it!

**Pull request merged**